storm-webapp/src/main/java/org/apache/storm/daemon/ui/UIHelpers.java

```diff
@@ -0,0 +1,1941 @@
+                spoutData.put("stats", spoutSummaries.get(
+                        spoutComponentId).stream().map(
+                                UIHelpers::getStatMapFromExecutorSummary).collect(Collectors.toList()));
+                spoutData.put("link", UIHelpers.urlFormat("/component.html?id=%s&topology_id=%s", spoutComponentId, topoId));
+
+                result.put("inputs",
+                    spoutSpecMapEntry.getValue().get_common().get_inputs().entrySet().stream().map(
+                            UIHelpers::getInputMap).collect(Collectors.toList())
+                );
+                result.put(spoutComponentId, spoutData);
+            }
+        }
+
+        for (Map.Entry<String, Bolt> boltEntry : boltSpecs.entrySet()) {
+            String boltComponentId = boltEntry.getKey();
+            if (boltSummaries.containsKey(boltComponentId) && (sys || Utils.isSystemId(boltComponentId))) {
+                Map<String, Object> boltMap = new HashMap();
+                boltMap.put("type", "bolt");
+                boltMap.put("capacity", StatsUtil.computeBoltCapacity(boltSummaries.get(boltComponentId)));
+                Map<String, Map> boltStreamsStats = StatsUtil.boltStreamsStats(boltSummaries.get(boltComponentId), true);
+                boltMap.put("latency", boltStreamsStats.get("process-latencies").get(window));
+                boltMap.put("transferred", boltStreamsStats.get("transferred").get(window));
+                boltMap.put("stats", boltSummaries.get(
+                        boltComponentId).stream().map(
+                        UIHelpers::getStatMapFromExecutorSummary).collect(Collectors.toList()));
+                boltMap.put("link", UIHelpers.urlFormat("/component.html?id=%s&topology_id=%s", boltComponentId, topoId));
+
+                result.put("inputs",
+                        boltEntry.getValue().get_common().get_inputs().entrySet().stream().map(
+                                UIHelpers::getInputMap).collect(Collectors.toList())
+                );
+                result.put(boltComponentId, boltMap);
+            }
+        }
+        return result;
+    }
+
+    /**
+     * getStreamBox.
+     * @param visualization visualization
+     * @return getStreamBox
+     */
+    public static Map<String, Object> getStreamBox(Object visualization) {
+        Map<String, Object> visualizationData = (Map<String, Object>) visualization;
+        Map<String, Object> result = new HashMap();
+        Map<String, Object> temp = (Map<String, Object>) visualizationData.get("inputs");
+        result.put("stream", temp.get("stream"));
+        result.put("sani-stream", temp.get("sani-stream"));
+        result.put("checked", !Utils.isSystemId((String) temp.get("stream")));
+        return result;
+    }
+
+    /**
+     * getBuildVisualization.
+     * @param client client
+     * @param config config
+     * @param window window
+     * @param id id
+     * @param sys sys
+     * @ret
```

Comment:
Not totally sure what happened here but the visualization does not work because the groupings are not being mapped to JSON correctly.

```
{"inputs":[{"component":"__acker","sani-stream":"_s__ack_reset_timeout","stream":"__ack_reset_timeout","grouping":<Grouping direct:NullStruct()>},{"component":"__acker","sani-stream":"_s__ack_ack","stream":"__ack_ack","grouping":<Grouping direct:NullStruct()>},{"component":"__acker","sani-stream":"_s__ack_fail","stream":"__ack_fail","grouping":<Grouping direct:NullStruct()>}],"spout":{"transferred":2800,"stats":[{"transferred":{"10800":{"default":2800},"600":{"default":2800},"86400":{"default":2800},":all-time":{"default":2800}},"port":6700,"host":"10.0.1.12","uptime_secs":61},{"transferred":{"10800":{"default":560},"600":{"default":560},"86400":{"default":560},":all-time":{"default":560}},"port":6701,"host":"10.0.1.12","uptime_secs":60},{"transferred":{"10800":{"default":560},"600":{"default":560},"86400":{"default":560},":all-time":{"default":560}},"port":6702,"host":"10.0.1.12","uptime_secs":61},{"transferred":{"10800":{"default":560},"600":{"default":560},"86400":{"default":560},":all-time":{"default":560}},"port":6700,"host":"10.0.1.12","uptime_secs":61},{"transferred":{"10800":{"default":560},"600":{"default":560},"86400":{"default":560},":all-time":{"default":560}},"port":6702,"host":"10.0.1.12","uptime_secs":61}],"latency":null,"link":"\/component.html?id=spout&topology_id=wc-1-1533304535","type":"spout","capacity":0}}
```

`<Grouping direct:NullStruct()>` is not what we want as output.